music21/metadata/__init__.py

```diff
@@ -142,115 +238,376 @@ class Metadata(base.Music21Object):
-        'oco': 'commission',
-        'ocy': 'countryOfComposition',
-        'ode': 'dedication',
-        'omd': 'movementName',
-        'omv': 'movementNumber',
-        'onm': 'number',
-        'opc': 'localeOfComposition',  # origin in abc
-        'opr': 'parentTitle',
-        'ops': 'opusNumber',
-        'osc': 'sceneNumber',
-        'ota': 'alternativeTitle',
-        'otl': 'title',
-        'otp': 'popularTitle',
-        'ovm': 'volume',
-        'txl': 'textLanguage',
-        'txo': 'textOriginalLanguage',
-    }
+    # We differentiate between allUniqueNames, a list of all the values you can get,
+    # and searchAttributes, a list of all the names you can search for.  The difference
+    # is that searchAttributes has all the extra workIds in it that are different from
+    # the uniqueName for that same property.  For example, 'dedicatedTo' is a uniqueName
+    # for which the workId is 'dedication'.  Both will show up in searchAttributes (so
+    # you can search for either one), but only 'dedicatedTo' will show up in allUniqueNames,
+    # because you don't want to get both 'dedicatedTo' and 'dedication', since they are
+    # the same value.  Note that searchAttributes' initialization uses set() to get rid
+    # of any workIds that are the same name as the uniqueName.
 
     # add more as properties/import exists
-    searchAttributes = tuple(sorted([
-        'composer',
-        'copyright',
-        'date',
+    allUniqueNames = tuple(sorted([
         'fileFormat',
         'fileNumber',
         'filePath',
-    ] + list(workIdAbbreviationDict.values())))
+    ] + properties.ALL_UNIQUENAMES))
 
-    workIdLookupDict = {}
-    for key, value in workIdAbbreviationDict.items():
-        workIdLookupDict[value.lower()] = key
+    searchAttributes: t.Tuple = tuple(sorted(set(
+        list(allUniqueNames) + properties.ALL_M21WORKIDS)))
 
     # INITIALIZER #
 
     def __init__(self, *args, **keywords):
         super().__init__()
 
-        # a list of Contributor objects
-        # there can be more than one composer, or any other combination
-        self.contributors = []  # use addContributor to add.
-        self._date = None
+        self._metadata: t.Dict = {}
+        self.software: t.List[str] = [defaults.software]
+
+        # TODO: check pickling, etc.
+        self.fileInfo = FileInfo()
+
+        # We allow the setting of metadata values (attribute-style) via **keywords.
+        # Any keywords that are uniqueNames, grandfathered workIds, or grandfathered
+        # workId abbreviations can be set this way.
+        for attr in keywords:
+            setattr(self, attr, keywords[attr])
 
-        # store one or more URLs from which this work came; this could
-        # be local file paths or otherwise
-        self._urls = []
+# -----------------------------------------------------------------------------
+
+    def add(self,
+            key: str,
+            value: t.Union[t.Any, t.Iterable[t.Any]]):
```

Comment:
Yeah, I can replace t.Any with t.Union[metadata.Text, metadata.DateSingle, metadata.Contributor, metadata.Copyright] through out.  I'm running out of Python knowledge at this point... is there a way to define a name for this in one place (say, metadata.ValueType) and then use that name in every API?  Then we can add more to it in one place, later, when we support more value types.

Reply:
Actually, `add()` and `__setitem__` and `__setattr__` _can_ all take t.Any, because there's an automatic conversion to the internal preferred valueType (which is Text, DateSingle, Contributor, or Copyright).  If the conversion fails, ValueError is raised.  That said, there are still places where t.Union[metadata.Text, ...] would be welcome for mypy checking.

Reply:
Fixed (using the new ValueType).